mypy/checker.py

```diff
@@ -186,6 +187,13 @@ def visit_file(self, file_node: MypyFile, path: str) -> None:
                 self.fail(messages.ALL_MUST_BE_SEQ_STR.format(str_seq_s, all_s),
                           all_.node)
 
+    @contextmanager
+    def new_scope(self) -> Iterator[ConditionalTypeBinder]:
+        old_binder = self.binder
+        self.binder = ConditionalTypeBinder()
```

Comment:
Won't this mean that we lose conditional type binder information about other variables in scope that might be used in the comprehension?

Also, in python 2 mode, it might be nice to mark the variables bound by a comprehension as deleted after the comprehension, since using them is likely a bug. Perhaps that is a separate issue though.